Add observer metric

krnowak opened the pull request:
Here is the initial stab at observers.

This adds integral and floating point observer instruments. The meter got `Register{Int,Float}64Observer(name, callback, options...)` functions. The observers are interfaces like as follows:

```go
type {Int,Float}64Observer interface {
	SetCallback(callback)
	Unregister()
}
```

The callback is `func({Int,Float}64ObserverResult)` and the result type is:

```go
type {Int,Float}64ObserverResult interface {
	Observe({int,float}64 value, LabelSet)
}
```

Both observer and observer result interfaces are implemented by SDK and mocks, noop and global only implement observer.

On the SDK side, each labelset passed to the result's observe function gets its own recorder/aggregator. So, every observer may have several aggregators if the callback called observe several times, each time with a different labelset. That more-or-less mimics what happens when you bind an instrument several times, each time with a different labelset.

There are also three new aggregators for observers, which are basically compositions of two other aggregators: (gauge, minmaxsumcount), (gauge, ddsketch) and (gauge, array). These are used by the simple selector for observer instruments.

SDK calls the observer callbacks in the collection routine, after checkpointing records. Currently there is no protection against a callback that does `sleep(time.Eternity)`.

TODO:

- ~~docs~~
  - ~~some are done, need more package docs (especially api/metric and sdk/metric)~~
- ~~tests~~
- ~~monotonicity option for observers~~
- ~~the `observe` implementation in sdk could likely be split into two parts:~~
  - ~~first part would remain in `observe` and could call Update on recorder every time observer callback calls `Observe` on result.~~
  - ~~second part would be moved to the observer loop in `Collect` and could call Process on batcher after observer callback is finished~~
- ~~the checkpointed variable should rather be incremented once per observer callback, not once per observation in the observer callback, I think~~
- ~~add observer aggregators and update aggregator selectors to handle the observer metric kind~~

Fixes #423

krnowak commented:
I must have some misunderstanding of the observers. During collection period we are going to call the callback of the observer. The callback receives a result object on which the callback can call the observe function many times. Also, observer is like a gauge, so by default we remember the value from the last event. So if we call the observe function several times within a single callback, we basically overwrite the last value all the time. Is this some corner case where calling observe multiple times doesn't make sense?

Also, I tried splitting the observe function as I described in the PR description and it looks to me that the value is not connected to the passed label set. We seem to care about label sets during checkpoint, for the batcher.

I wonder if it would make sense for the observer result object to have two functions:

```go
type Int64ObserverResult interface {
    Observe(value ...int64)
    Labels(labels ...LabelSet)
}
```

Anyway, I need to read and understand the [standard implementation for observers](https://github.com/open-telemetry/opentelemetry-specification/blob/dc61b4eb4b95dfcc9e4ab756d5cd6c8d0c21aa66/specification/api-metrics.md#standard-implementation) first. This seems to invalidate a lot I wrote above.

krnowak commented:
Scratch the above post. Now each observer has a separate recorder for every labelset passed to the observe function and it makes more sense now.

krnowak commented:
I'd like to get some initial opinion on the PR before I embark on a dull task of writing the tests for observers. ;)

mauriciovasquezbernal commented on api/metric/api.go:
```diff
@@ -127,10 +135,51 @@ type Meter interface {
+	Observe(value int64, labels LabelSet)
+}
+
+// Int64ObserverResult is an interface for reporting floating point
```
```suggestion
// FloatObserverResult is an interface for reporting floating point
```

krnowak replied:
Thanks for spotting. :)

rghetia commented on api/global/internal/meter.go:
```diff
@@ -203,6 +237,121 @@ func (bound *instHandle) Unbind() {
+func (obs int64ObsImpl) SetCallback(callback metric.Int64ObserverCallback) {
+	if obs.trySetCallbackInDelegate(callback) {
+		return
+	}
```
Are two attempts necessary? 
Wouldn't  this work?
```
obs.observer.setCallback(callback)
obs.trySetCallbackInDelegate(callback)
```

krnowak replied:
That would work too, but it is an (unprofiled ;) ) optimization - `obs.observer.setCallback(callback)` involves locking the mutex. And I expect the first call to `trySetCallbackInDelegate` to usually succeed anyway.

rghetia commented on api/metric/api.go:
```diff
@@ -127,10 +135,53 @@ type Meter interface {
 	// a given name and customized with passed options.
 	NewFloat64Measure(name string, mos ...MeasureOptionApplier) Float64Measure
 
+	// NewInt64Observer creates a new integral observer with a
```
nit:s/NewInt64Observer/RegisterInt64Observer/

krnowak replied:
Done.

rghetia commented on api/metric/api.go:
```diff
@@ -127,10 +135,53 @@ type Meter interface {
+	// given name, running a given callback, and customized with
+	// passed options. Callback can be nil.
+	RegisterInt64Observer(name string, callback Int64ObserverCallback, oos ...ObserverOptionApplier) Int64Observer
+	// NewFloat64Observer creates a new floating point observer
```
same.

krnowak replied:
Done.

rghetia commented on api/metric/api.go:
```diff
@@ -244,16 +316,19 @@ func WithKeys(keys ...core.Key) OptionApplier {
-func WithMonotonic(monotonic bool) CounterGaugeOptionApplier {
-	return counterGaugeOptionWrapper{
+// WithMonotonic sets whether a counter, a gauge or an observer is not
+// permitted to go down.
```
May be we should note that this option still requires some work in the spec and should be used with caution.
For example, how will observer monotonicity will be enforced?

jmacd replied:
This is for a future PR.

jmacd commented on sdk/metric/sdk.go:
```diff
@@ -115,16 +119,141 @@ type (
+		descriptor *export.Descriptor
+		// recorders maps encoded labelset to the pair of
+		// labelset and recorder
+		recorders map[string]labeledRecorder
```
This looks like it could be a source of unbounded memory growth. We need to be clear about when it is cleared.

Two possible opinions:
(1) We shouldn't need to store this map in the observer at all. We only need to de-duplicate a particular observation if another appearance of the same label set has been observed in the same interval. This means we only need such a map to be built up during collection, not persisted across collections. This would lead to new Aggregators being allocated for each observation.
(2) We should only need to store the map of observers that were used in the last collection interval, that way we can re-use their Aggregator object and avoid an allocation. However, we should be sure that if a label set stops being observed, the memory is cleared. However, we don't have an Aggregator API to "reset" at present, and I think that would be needed, since if we re-use an Aggregator across intervals we need a way to zero it out.

krnowak replied:
As mentioned during the meeting - we remove an aggregator for a specific label set if the label set hasn't got any observation for two collection intervals in a row. If the aggregator map becomes empty, we drop the map too - see the [`checkpointObserver`](https://github.com/open-telemetry/opentelemetry-go/blob/0972daae77faa6bf6a37823fbca1422cbfdae8f6/sdk/metric/sdk.go#L590-L611) function. I admit that "two collection intervals" is arbitrary. Maybe we should simply state that if an aggregator/labelset didn't get any observation in the collection interval, it gets dropped. This is assuming that observation callbacks should be more or less a simple sequence of calls to `result.Observe(…)` (that is - no conditional observations).

Short notes about opinions above:
Ad (1): I can imagine having two observers with callbacks, where both are using the same label sets, and a custom aggregator selector that assigns minmaxsumcount aggregator to one observer and some last value stuff to the second observer. That's probably a fringe case, but in such a case we would have two different aggregators for the same label set, thus having a behavior depending on the order of iteration of the `observers` map.

Ad (2): That's quite similar to the current solution, but instead of a map of observers used in last collection interval we have a map (`map[encodedLabelSet]Aggregator`) in each observer. Entries are dropped from the map when no observation happens, so if the label set receives an observation again, we create an Aggregator again - this is a way of "zeroing" the aggregator.

krnowak commented on api/metric/api.go:
```diff
@@ -168,16 +222,22 @@ type measureOptionWrapper struct {
-	_ MeasureOptionApplier = measureOptionWrapper{}
+	_ OptionApplier         = optionWrapper{}
+	_ CounterOptionApplier  = counterOptionWrapper{}
+	_ GaugeOptionApplier    = gaugeOptionWrapper{}
```
We have decided during the SIG meeting to drop Gauge after adding Observer.

krnowak commented on sdk/metric/aggregator/observerarray/observerarray.go:
```diff
@@ -0,0 +1,115 @@
+// Copyright 2020, OpenTelemetry Authors
```
That's fine - less code means fewer tests too. ;)

krnowak commented:
Updated.

krnowak commented:
TestDirect seems to fail for some reason. I have run it on my computer in 100 iterations loop and all passed. :/

I suppose that we should be checking the core.Number equality in a different way, especially when dealing with floats.

krnowak commented:
Rebased. And fixed the `TestDirect` failure. Basically global meter was registering the observers at the delegate meter in random order. And mock meter was calling observer callbacks in random order too, so the order of taken measurements was also random. `TestDirect` was expecting some fixed order. My earlier runs of 100 iterations were badly done, because `go test` cached the test results. Go me.

Added some tests for the internal global meter alignment stuff - test 386 was failing on those.

jmacd commented on sdk/metric/sdk.go:
```diff
@@ -115,16 +119,143 @@ type (
+	if o.recorders == nil {
+		o.recorders = make(map[string]labeledRecorder)
+	}
+	// TODO: This may store nil recorder in the map, thus
```
This is intentionally done for the other instruments. We want a way to disable these, but you could be right about this: re-querying the selector API allows us to dynamically reconfigure it later. We're so behind on configuration work I'd be happy to enter a comment here saying this is intentional and revisit it later.

krnowak replied:
I updated the comment then.

jmacd commented:
One thing I wonder about is why we should have a `SetCallback` interface, as opposed to a constructor argument that would bind the callback once for the lifetime of the Observer?

@bogdandrutu It occurs to me that we haven't specified this detail, what do you think?

rghetia commented:
> One thing I wonder about is why we should have a `SetCallback` interface, as opposed to a constructor argument that would bind the callback once for the lifetime of the Observer?

I agree. Observer without a Callback is incomplete and most likely it would not require changing it. If we do find a use case later to update the callback then we can always add the SetCallback or UpdateCallback api.

MrAlias commented on sdk/metric/selector/simple/simple.go:
```diff
@@ -38,30 +38,30 @@ var (
-// kinds of metric.  This selector is faster and uses less memory than
-// the others because minmaxsumcount does not aggregate quantile
-// information.
+// that uses counter, gauge, minmaxsumcount, and observermmsc
```
Just saw the history of the PR, looks like these are indeed holdovers from a previous idea. +1 on removing here.

krnowak replied:
Yeah, old stuff I did not clean up properly.

MrAlias commented:
> One thing I wonder about is why we should have a SetCallback interface, as opposed to a constructor argument that would bind the callback once for the lifetime of the Observer?

The `Register*Observer` functions seem to meet the constructor binding approach, no?

The one thing I would be worried about with removing the `SetCallback` interface is if the instrumentor had to deal with a changing object they wanted to instrument. Like a references to an ephemeral worker queue or an ad hoc stack. If they were wrapping the callback around these objects (admittedly, probably a bad idea on their part), they would want to update the callback when the references changed.

krnowak commented on api/global/internal/meter.go:
```diff
@@ -296,3 +457,64 @@ func (m *meter) NewInt64Measure(name string, opts ...metric.MeasureOptionApplier
+	m.observersOrdered = append(m.observersOrdered, obs)
+}
+
+func AtomicFieldOffsets() map[string]uintptr {
```
Tests are a separate package (`internal` vs `internal_test`), so I can't access the private fields of the private struct of the `internal` package in tests. So I could either rename the tests package to `internal`, or export the structs and their members, or write this function. I went with the last one. :)

MrAlias replied:
Ah! Correct, thanks for pointing it out 😄

krnowak commented on api/global/internal/meter.go:
```diff
@@ -40,44 +40,78 @@ const (
+	// registration. It may also contain unregistered
+	// observers. The liveObservers map should be consulted to
+	// check if the observer is registered or not.
+	observersOrdered []*obsImpl
```
In the beginning it was just `observers` and `observersOrdered`, then went through a bunch of refactors and renamings. Thanks for the suggestion, will do.

krnowak commented on api/metric/api.go:
```diff
@@ -127,10 +135,53 @@ type Meter interface {
 	RecordBatch(context.Context, LabelSet, ...Measurement)
 }
 
+// Int64ObserverResult is an interface for reporting integral
```
Isn't "integral" an adjective form of "integer"? I have used that in other docs too.

MrAlias replied:
Wow! I have always used it to mean the function opposite of a derivative or the area under a curve. TIL.

krnowak commented on sdk/metric/doc.go:
```diff
@@ -61,31 +73,18 @@ from the user's perspective.
+epoch during which it was updated.  This variable allow the collection code
+path to detect stale recorders and remove them.
+
+Each record and recorder in observer has an associated aggregator, which
```
Yeah, I think this is something we will need to review too - recorder and aggregator seem to be used interchangeably. Here I reworded it for now to `Each record of a handle and recorder of an observer…`.

krnowak commented on sdk/metric/sdk.go:
```diff
@@ -358,25 +557,68 @@ func (m *SDK) Collect(ctx context.Context) int {
-	labels := export.NewLabels(r.labels.sorted, r.labels.encoded, m.labelEncoder)
-	err := m.batcher.Process(ctx, export.NewRecord(r.descriptor, labels, r.recorder))
+	checkpointed := 0
+	ctx := context.Background()
```
No clue what I was thinking when I wrote it - too much time has passed since then. :P Probably got carried away with thinking about observers differing from other instruments, that observer callbacks lack the context other instruments have at the time of recording the measurement.

But you are right, will fix it.

krnowak commented on sdk/metric/selector/simple/simple.go:
```diff
@@ -38,30 +38,30 @@ var (
-// of metric.  This selector uses more cpu and memory than the
-// NewWithInexpensiveMeasure because it uses one DDSketch per distinct
-// measure and labelset.
+// uses counter, gauge, ddsketch, and observerddsketch aggregators for
```
Same, removed it.

krnowak commented on sdk/metric/selector/simple/simple.go:
```diff
@@ -38,30 +38,30 @@ var (
-// This selector uses more memory than the NewWithSketchMeasure
-// because it aggregates an array of all values, therefore is able to
-// compute exact quantiles.
+// counter, gauge, array, and observerarray aggregators for the four
```
Same, removed it.

krnowak commented:
Rebased to master. Dropped the SetCallback function.

jmacd commented:
:clap:
